Adapt to Erlang/OTP 23

paulo-ferraz-oliveira opened the pull request.

paulo-ferraz-oliveira committed:
Adapt http_uri:decode to OTP 23

paulo-ferraz-oliveira committed:
Fix broken tests in OTP 23

paulo-ferraz-oliveira committed:
Brush up rebar.config a bit

Also fixes `rebar edoc` and makes for faster dialyzer'ing

paulo-ferraz-oliveira committed:
Fix broken tests in OTP 19

paulo-ferraz-oliveira committed:
Improve test-based consumer confidence

paulo-ferraz-oliveira committed:
Prepare for future release

paulo-ferraz-oliveira committed:
Try to trigger Travis CI

g-andrade commented on .travis.yml:
```diff
@@ -1,14 +1,19 @@
+   - 22.0
+   - 22.1
+   - 22.2
+   - 22.3
```
^ We should add a `23.0` target here, given the objective of this PR.

paulo-ferraz-oliveira replied:
I searched for it, couldn't find it. Where's the canonical list?

g-andrade replied:
There isn't one, I think.

For new versions, I usually [poll google](https://www.google.com/search?q=erlang+23+travis) until [something](https://travis-ci.community/t/erlang-23-0-released/8606) pops out confirming there is an image available.

paulo-ferraz-oliveira replied:
I'd found https://docs.travis-ci.com/user/languages/erlang/#choosing-otp-releases-to-test-against and it seems to point to `kerl`, so I'm simply adding 23.0 as per your recommendation (then we'll check the results, after Travis CI is available).

g-andrade commented on src/lhttpc_lib.erl:
```diff
@@ -239,12 +239,26 @@ split_credentials(CredsHostPortPath) ->
 
+-ifdef(OTP_RELEASE).
+-if(?OTP_RELEASE >= 23).
+http_uri_decode(User0) ->
```
I would prefer we name it something else other than `http_uri:decode/1`, since it doesn't actually decode an URI, but rather it decomposes/dissects/parses a query string.

paulo-ferraz-oliveira replied:
What name do you propose? I'll use it as you mention it. I want to waste no more time on this PR.

g-andrade replied:
`parse_query_string`?

paulo-ferraz-oliveira replied:
Cool, thx, bye.

g-andrade reviewed (changes requested):
The behaviour of TLS certificate verification in [ssl](https://erlang.org/doc/man/ssl.html) in OTP 23 changed due to a [bugfix](https://bugs.erlang.org/browse/ERL-1232) and this brought to light some issues with wildcard certificates.

OTP 22:
```
% lhttpc:request("https://download.maxmind.com/", "HEAD", [], <<>>, 5000).
{ok,{{301,"Moved Permanently"}, % ...
 }}
```

OTP 23:
```
% lhttpc:request("https://download.maxmind.com/", "HEAD", [], <<>>, 5000).
{error,{{tls_alert,{handshake_failure,"TLS client: In state wait_cert_cr at ssl_handshake.erl:1840 generated CLIENT ALERT: Fatal - Handshake Failure\n {bad_cert,hostname_check_failed}"}},
        [{lhttpc_client,send_request,1,
                        [{file,"/home/guilherme/code/server/lhttpc/src/lhttpc_client.erl"},
                         {line,242}]},
         {lhttpc_client,execute,9,
                        [{file,"/home/guilherme/code/server/lhttpc/src/lhttpc_client.erl"},
                         {line,187}]},
         {lhttpc_client,request,9,
                        [{file,"/home/guilherme/code/server/lhttpc/src/lhttpc_client.erl"},
                         {line,95}]}]}}
```

^ Given that the changes required to fix this issue are not compatible with OTP 19 (requiring even more ifdef'ing), I'm propose we use [`tls_certificate_check`](https://hex.pm/packages/tls_certificate_check) instead and drop the direct integration of `certifi` and `ssl_verify_fun`.

paulo-ferraz-oliveira commented:
> The behaviour of TLS certificate verification in [ssl](https://erlang.org/doc/man/ssl.html) in OTP 23 changed due to a [bugfix](https://bugs.erlang.org/browse/ERL-1232) and this brought to light some issues with wildcard certificates.
> 
> OTP 22:
> 
> ```
> % lhttpc:request("https://download.maxmind.com/", "HEAD", [], <<>>, 5000).
> {ok,{{301,"Moved Permanently"}, % ...
>  }}
> ```
> 
> OTP 23:
> 
> ```
> % lhttpc:request("https://download.maxmind.com/", "HEAD", [], <<>>, 5000).
> {error,{{tls_alert,{handshake_failure,"TLS client: In state wait_cert_cr at ssl_handshake.erl:1840 generated CLIENT ALERT: Fatal - Handshake Failure\n {bad_cert,hostname_check_failed}"}},
>         [{lhttpc_client,send_request,1,
>                         [{file,"/home/guilherme/code/server/lhttpc/src/lhttpc_client.erl"},
>                          {line,242}]},
>          {lhttpc_client,execute,9,
>                         [{file,"/home/guilherme/code/server/lhttpc/src/lhttpc_client.erl"},
>                          {line,187}]},
>          {lhttpc_client,request,9,
>                         [{file,"/home/guilherme/code/server/lhttpc/src/lhttpc_client.erl"},
>                          {line,95}]}]}}
> ```
> 
> ^ Given that the changes required to fix this issue are not compatible with OTP 19 (requiring even more ifdef'ing), I'm propose we use [`tls_certificate_check`](https://hex.pm/packages/tls_certificate_check) instead and drop the direct integration of `certifi` and `ssl_verify_fun`.

Which means tests aren't covering it, right? What do you propose tests should bring to light, before I import that lib. of yours?

g-andrade commented:
> Which means tests aren't covering it, right? What do you propose tests should bring to light, before I import that lib. of yours?

There aren't any. At least, not without implementing a whole local SSL service with lots of mocking in between. (The lib. itself doesn't yet cover wildcard certificates in its tests, either.)

We can do a PR to [badssl](https://github.com/chromium/badssl.com), which we currently use in the tests, but I'm not sure they would accept something "as basic" as checking for wildcard certificate support.

g-andrade commented:
In any case, if we do choose to use `tls_certificate_check`, I propose removing the test cases for SSL certificate verification from `lhttpc`, since they're pretty much the same as `tls_certificate_check`'s.

paulo-ferraz-oliveira committed:
Adapt Travis CI targets to our reality

I still have no idea where the canonical target list is found
A previous "list" I'd found didn't contain (still doesn't) OTP 23

paulo-ferraz-oliveira committed:
Rename for clarity

paulo-ferraz-oliveira commented:
> In any case, if we do choose to use `tls_certificate_check`, I propose removing the test cases for SSL certificate verification from `lhttpc`, since they're pretty much the same as `tls_certificate_check`'s.

Can you identify those? I don't mind doing it, but have no idea what test cases we're talking about.

paulo-ferraz-oliveira commented:
And, yes, we are adopting `tls_certificate_check`.

paulo-ferraz-oliveira committed:
Ease the establishement of more secure HTTPS connections

paulo-ferraz-oliveira commented:
> And, yes, we are adopting `tls_certificate_check`.

This has happened... I didn't update the test suites, yet, but tests are passing, so don't know if it requires handling.

paulo-ferraz-oliveira committed:
Bow to Travis CI's cruelty

paulo-ferraz-oliveira committed:
Don't repeat ourselves, by removing test coverage for tls_certificate_check-covered stuff

These are not failing, though, it's just a simple trim down. Coverage went from 70% to 71%

paulo-ferraz-oliveira committed:
Allow bypassing Travis CI vs. localdev limitations

e.g. no support for IPv6

g-andrade commented:
> Can you identify those? I don't mind doing it, but have no idea what test cases we're talking about.

[These ones](https://github.com/MiniclipPortugal/lhttpc/blob/3.2.0/test/lhttpc_test_SUITE.erl#L800-L822).

g-andrade commented:
Ohh, I now realize you already found them.

paulo-ferraz-oliveira committed:
Try jumping over lacking IPv6 support

paulo-ferraz-oliveira commented:
I'm pushing changes that allow skipping IPv6 -related tests (it seems Travis CI is lacking support for it).

paulo-ferraz-oliveira committed:
Bypass internal dialyzer/OTP issue on missing public_key type

paulo-ferraz-oliveira committed:
Comment out Travis CI version for which tests are failing with a timeout

This smells likes an OTP issue, but I don't want to deal with it

paulo-ferraz-oliveira committed:
Reverse unwarranted white space changes
